src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -1827,23 +1827,25 @@ private DisposePluginService SearchAndInitializeProjectCachePluginFromGraph(Proj
                 return new DisposePluginService(null);
             }
 
-            var cacheItems = projectGraph.ProjectNodes
-                .Select(n => n.ProjectInstance)
-                .SelectMany(p => p.GetItems(ItemTypeNames.ProjectCachePlugin))
-                .Select(
-                    i =>
-                    {
-                        var metadataDictionary = i.Metadata.ToDictionary(
-                            m => ((IKeyed) m).Key,
-                            m => ((IValued) m).EscapedValue);
+            var nodeToCacheItems = projectGraph.ProjectNodes.ToDictionary(
+                n => n,
+                n => n.ProjectInstance.GetItems(ItemTypeNames.ProjectCachePlugin)
+                    .Select(
+                        i =>
+                        {
+                            var metadataDictionary = i.Metadata.ToDictionary(
+                                m => ((IKeyed) m).Key,
+                                m => ((IValued) m).EscapedValue);
 
-                        var pluginPath = Path.Combine(i.Project.Directory, i.EvaluatedInclude);
+                            var pluginPath = Path.Combine(i.Project.Directory, i.EvaluatedInclude);
 
-                        var projectCacheItem = new ProjectCacheItem(pluginPath, metadataDictionary);
+                            var projectCacheItem = new ProjectCacheItem(pluginPath, metadataDictionary);
 
-                        return projectCacheItem;
-                    })
-                .ToHashSet();
+                            return projectCacheItem;
+                        })
+                    .ToArray());
+
+            var cacheItems = nodeToCacheItems.Values.SelectMany(i => i).ToHashSet();
```

Comment:
What's the purpose of `SelectMany` here?

Reply:
I'm not sure I follow why the conversion to an array then a hashset, can't the list of items be stored as a hashset to begin with?

Reply:
It's inconsequential, I think. Each node can declare zero or more plugins, so first I construct a dictionary from a node to the collection of plugins it declares. But the contract is that there can be a single plugin (path + plugin settings), and all nodes must declare that plugin. So in order to find that single plugin I flatten (SelectMany) the collections of plugins from each node into a single set (ProjectCacheItem implements Equals and GetHashCode to make this correct). I could have skipped the dictionary by flattening everything from the start, but I also want to give a nice error message with all the nodes that may be missing the plugin. And given that this method is not on a hot path, I went with a lot of LINQ :)

Reply:
I'm not sure I follow everything here. Wouldn't this be union rather than intersection, so wouldn't this find all the plugins declared by any node? How does this help you find the one from all of them? And I'm not sure how I follow how this turns into an error.

Reply:
This part constructs a dictionary from every node to all the declared plugins in that node:
https://github.com/cdmihai/msbuild/blob/acb4b7e2e0ffbd4a9df13c6fba11dd6aa5f37944/src/Build/BackEnd/BuildManager/BuildManager.cs#L1866-L1882

This part flattens (via SelectMany) all the declared plugins into a single set which removes the duplicates according to the overriden equals and hashcode in [ProjectCacheItem](https://github.com/cdmihai/msbuild/blob/acb4b7e2e0ffbd4a9df13c6fba11dd6aa5f37944/src/Build/BackEnd/Components/ProjectCache/ProjectCacheItem.cs#L29-L69).
https://github.com/cdmihai/msbuild/blob/acb4b7e2e0ffbd4a9df13c6fba11dd6aa5f37944/src/Build/BackEnd/BuildManager/BuildManager.cs#L1884

The set should contain a single item if all nodes declare a single plugin (plugin path + plugin settings).

Error cases:
- set contains more than 1 plugins (error prints all declared plugins)
- set contains a single plugin but there are nodes which do not declare it (error prints all nodes that do not declare the plugin)